comprobo_final_project/scripts/tasks/goal_task.py

```diff
@@ -0,0 +1,135 @@
+import time
+import numpy as np
+from ..models.robot import Robot as ModelRobot
+from ..simulator.robot import Robot as SimRobot
+from ..simulator.simulation_visualizer import SimulationVisualizer
+from ..gene_alg_2.genetic_algorithm import GeneticAlgorithm
+from ..visualizations import fitness_vs_run
+
+
+class GoalTask(object):
+    """
+    Allows a robot to move to a fixed goal.
+    """
+
+    def train(self, robot):
+        """
+        Trains the task to find the most fit organism. Organisms are stored in a
+        log file along with their fitness per generation.
+        """
+
+        log_location = 'logs/log_'+str(int(time.time()))+'.csv'
+        print log_location
+
+        GeneticAlgorithm(
+            log_location=log_location,
+            gen_size=100,
+            num_genes=4,
+            elitism_thresh=0.1,
+            crossover_thresh=0.8,
+            mutation_thresh=0.5,
+            fitness_func=self.get_fitness_func(robot)).train()
+
+
+    def visualizer_test(self, robot, organism):
+        """
+        Use the basic visualizer to see what the robot is doing.
+        """
+        simulation_visualizer = SimulationVisualizer(robot, real_world_scale=2)
+        self.run_with_setup(robot, organism)
+
+
+    def get_fitness_func(self, robot):
+        """
+        Provides a fitness function for the genetic alorigthm to optimize for.
+        """
+
+        # Using a closure here so we can hold our single robot instance
+        def _fitness_func(organism):
+            positions = self.run_with_setup(robot, organism)
+
+            distances = [np.sqrt(position.x**2 + position.y**2) \
+                    for position in positions] # all distances from goal
+            fitness = np.mean(distances) # average distance from goal
+
+            return fitness
+
+        return _fitness_func
```

Comment:
yep, you define a function, and then return it. This creates a closure, which allows that function to having a binding to our `robot`. This way, we don't have to pass our `robot` into the GeneticAlgorithm, which would be wrong.